File: Dept3 面向对象程序设计/Duck.py
# 问题：本例子使用了鸭子🦆类型，多态性具体体现在哪里？
# Python 示例代码
class Duck:
    def quack(self):
        print("Quack")

    def walk(self):
        print("Walk")

class Person:
    def quack(self):
        print("I'm quacking like a duck")

    def walk(self):
        print("I'm walking like a duck")

def make_it_quack(obj):
    #
    obj.quack()


# Client 创建类对象    
duck = Duck() 
person = Person()


make_it_quack(duck)  # 输出：Quack
make_it_quack(person)  # 输出：I'm quacking like a duck

File: Dept3 面向对象程序设计/Duck.c
// C 示例代码
#include <stdio.h>

// 定义鸭子类型的函数
void make_it_quack(void (*quack)()) {
    quack();
}

// 定义鸭子对象
void duck_quack() {
    printf("Quack\n");
}

// 定义人类对象
void person_quack() {
    printf("I'm quacking like a duck\n");
}

int main() {
    // 传递鸭子对象
    make_it_quack(duck_quack);  // 输出：Quack
    
    // 传递人类对象
    make_it_quack(person_quack);  // 输出：I'm quacking like a duck
    
    return 0;
}
